### ISSUE TITLE
Pane titles

### ISSUE THREAD

**Issue by kond3** [tmux/tmux#4245]
Let's start saying that `tmux` is amazing, I discovered it about a year ago and quite immediately I started customizing my configuration. Now I'm stuck, I can't edit my `.tmux.conf` so that `tmux` displays a default name for panes such as `Pane <n>` and optionally get prompted to change that name to something I want.
I understand the scope of the `set pane-border-format` statement is actually the window and my current `.tmux.conf` is really close to what I want, my theory is that the way I used some statements is buggy and this is why I get an unexpected behavior, but I don't get why `pane-border-format` is a window property.

Interested lines in my `.tmux.conf`:

```
set -g pane-border-status top
set -g pane-border-format "[#{?pane_active,#[fg=colour49 bold],} pane #P ]#[fg=colour39,nobold]"
bind-key -n M-R command-prompt -p "Pane name:" 'select-pane -T "%%"' \; set pane-border-format "[#{?pane_active,#[fg=colour49 bold],} #T ]#[fg=colour39,nobold]"
```

Default pane title works just fine, I get a pretty layout when opening a new `tmux` session and splitting the window.

```
-----[ pane 0 ]---------------- | ----[ pane 1 ]----------------
                                |
                                |
```

When I hit Alt-R (the key bound to the prompt command in order to set a new title) all panes names default to `$USER`@`HOST`:`current-directory`, but when I enter a name it gets properly assigned to the `#{pane_title}` and displayed as in my `pane-border-format` directive.

```
-----[ asdf ]---------------- | ----[ $USER@$HOST:current-directory ]----------------
                              |
                              |
```

Now the title stays there, all other panes changes this value too and it's fine, I guess if I'm setting a new name for one pane, I will rename also the others in the same window, for example I'm debugging a web app and I want to make requests on a pane named "Requests" and read responses on a pane named "Responses". Problem arises now: if I enter a command, any command, pane name defaults again, and not to the custom `pane #P` format.

```
-----[ $USER@$HOST:current-directory ]---------- | ----[ $USER@$HOST:current-directory ]----------
ls /                                             |
bin  boot  cdrom  dev  etc  ...                  |
```

Long story short: is there a reason for `pane-border-format` to be a window property and not a pane one? Is it possible to achieve my ideal configuration using a different approach?

---

### For `CONTRIBUTING.md`
Output of `uname -sp && tmux -V && echo $TERM` is
```
Linux x86_64
tmux 3.2a
xterm-256color
```

I'm using `zsh` in Ubuntu 22.04, testing with `bash` I get the same results.

**Comment by nicm** [comment]
Why are you setting the option again when you rename a pane? You don't need to do that and since the `#T` is being replaced immediately all the panes become the same. If you must do it for some reason you will need to do `##T`.

The reason they change back is because your shell is changing the title, either configure it not to or turn `allow-set-title` to off.

**Comment by kond3** [comment]
I am setting the option again because I want the default to display the pane index (#P), and if I only change the title with Alt-R, this does not appear on the pane border. In order to see it I need to change the `pane-border-format`.
Adding the line
```
set-option -g allow-set-title off
```

**Comment by nicm** [comment]
Show server log please with `allow-set-titles off` and `##T` rather than `#T`,

**Comment by kond3** [comment]
Adding `##T` results in all panes being named "#T" when hitting `Alt-R`, I guess the double `##` does work as expected. Running `tmux -vv` Results in error `.tmux.conf:88: invalid option: allow-set-title`

`.tmux.conf`
![image](https://github.com/user-attachments/assets/437bdd59-1c61-43d1-b14e-55231012fcf7)

Hitting `Alt-R`
![image](https://github.com/user-attachments/assets/894b4d2b-9cfc-4a18-b2a7-1b32847ea908)

Error returned
![image](https://github.com/user-attachments/assets/75e1284e-d512-4e3f-82d3-7d93eb42e91e)

Interesting lines from `tmux-server-xxxxx.log`:
- When changing name with `Alt-R`
```
1732205660.655897 message: /dev/pts/1 key M-R: command-prompt -p "Pane name:" "select-pane -T \\"%%\\""
1732205660.655907 cmdq_fire_command </dev/pts/1>: (1703) command-prompt -p "Pane name:" "select-pane -T \\"%%\\""
1732205660.655912 cmd_find_from_client: s=$0 0
1732205660.655916 cmd_find_from_client: wl=0 1 w=@0 zsh
1732205660.655920 cmd_find_from_client: wp=%0
1732205660.655923 cmd_find_from_client: idx=none
1732205660.655928 cmd_find_from_client: s=$0 0
1732205660.655933 cmd_find_from_client: wl=0 1 w=@0 zsh
1732205660.655938 cmd_find_from_client: wp=%0
1732205660.655943 cmd_find_from_client: idx=none
1732205660.655951 format_defaults: c=/dev/pts/1
1732205660.655957 format_defaults: s=$0
1732205660.655960 format_defaults: wl=0
1732205660.655963 format_defaults: wp=%0
1732205660.655972 format_expand1: expanding format: Pane name: 
1732205660.655977 format_expand1: result is: Pane name: 
1732205660.655982 unref client 0x5aaaa351a650 (3 references)
1732205660.655986 cmdq_next </dev/pts/1>: [set-option/0x5aaaa35101c0] (0), flags 0
1732205660.656002 message: /dev/pts/1 key M-R: set-option -g pane-border-format "[#{?pane_active,#[fg=colour49 bold],} #T ]#[fg=colour39,nobold]"
1732205660.656013 cmdq_fire_command </dev/pts/1>: (1703) set-option -g pane-border-format "[#{?pane_active,#[fg=colour49 bold],} #T ]#[fg=colour39,nobold]"
1732205660.656017 cmd_find_from_client: s=$0 0
1732205660.656021 cmd_find_from_client: wl=0 1 w=@0 zsh
1732205660.656024 cmd_find_from_client: wp=%0
1732205660.656028 cmd_find_from_client: idx=none
1732205660.656034 cmd_find_target: target none, type pane, item 0x5aaaa35101c0, flags QUIET,CANFAIL
1732205660.656038 cmd_find_target: current is from queue
1732205660.656042 cmd_find_target: s=$0 0
1732205660.656046 cmd_find_target: wl=0 1 w=@0 zsh
1732205660.656050 cmd_find_target: wp=%0
1732205660.656053 cmd_find_target: idx=none
1732205660.656058 format_defaults: c=/dev/pts/1
1732205660.656062 format_defaults: s=$0
1732205660.656065 format_defaults: wl=0
1732205660.656068 format_defaults: wp=%0
1732205660.656072 format_expand1: expanding format: pane-border-format
1732205660.656077 format_expand1: result is: pane-border-format
```
- When entering a command such as `ls`
```
1732205679.052830 format_expand1: expanding format: [#{?pane_active,#[fg=colour49 bold],} #T ]#[fg=colour39,nobold]
1732205679.052835 format_expand1: found #{}: ?pane_active,#[fg=colour49 bold],
1732205679.052840 format_replace: condition is: pane_active
1732205679.052850 format_replace: condition 'pane_active' found
1732205679.052854 format_replace: condition 'pane_active' is true
1732205679.052858 format_expand1: expanding format: #[fg=colour49 bold]
1732205679.052862 format_expand1: result is: #[fg=colour49 bold]
1732205679.052868 format_replace: replaced '?pane_active,#[fg=colour49 bold],' with '#[fg=colour49 bold]'
1732205679.052872 format_expand1: found #T: pane_title
1732205679.052881 format_replace: format 'pane_title' found: $USER@$HOST:~/tmp
1732205679.052885 format_replace: replaced 'pane_title' with '$USER@$HOST:~/tmp'
1732205679.052892 format_expand1: result is: [#[fg=colour49 bold] $USER@$HOST:~/tmp ]#[fg=colour39,nobold]
1732205679.052898 screen_write_start: size 232x1, no pane
1732205679.052947 screen_write_cursormove: from 231,0 to 0,0
1732205679.052952 format_draw: [#[fg=colour49 bold] $USER@$HOST:~/tmp ]#[fg=colour39,nobold]
```

### Note
In log file `$USER` and `$HOST` were replaces by their actual values.

**Comment by nicm** [comment]
Your tmux is too old for `allow-set-title`, try upgrading to tmux 3.5a, or configure your shell not to change the title.

**Comment by nicm** [comment]
No there is no way to change the default title, except to change the key bindings for `split-window` to add `-T`.

**Comment by kond3** [comment]
Tried that and (it works but) got a strange behavior: when splitting from `pane 0`, also if there is another `pane 2` somewhere in the window, new pane gets named `pane 1`
![image](https://github.com/user-attachments/assets/0ab23ccc-96ec-44ef-b295-4bb98d98e58b)

In `.tmux.conf`
![image](https://github.com/user-attachments/assets/c4bf49e5-0f9d-4bf6-bbb4-ce2aadcb3a35)

**Comment by nicm** [comment]
Pane numbers are not static so yeah it might not be the same. You might be better to do it in `pane-border-format` instead, so leave the default title but have something like `#{?#{==:#{pane_title},#{host}},Pane #{pane_index},#{pane_title}}`. Or set the default title to something that can't change like `#{pane_id}` or the time.

**Comment by kond3** [comment]
Yeah this works too (there still is a slight "misconfiguration" since after splitting the first pane, going back to `pane 0` and splitting again, new pane becomes `pane 1` and pane that was `pane 1` becomes `pane 2`, I guess it's ok), but when renaming a pane, all the the others get named `$HOST`, instead of remaining `pane #P` (or `#{pane_index}`).

**Comment by nicm** [comment]
That's expected, pane indexes are not fixed, if you want fixed you will need to use pane ID.
